packages/network-controller/src/types.ts

```diff
@@ -30,6 +30,7 @@ export type CustomNetworkClientConfiguration = {
  * network.
  */
 export type InfuraNetworkClientConfiguration = {
+  chainId: Hex;
```

Comment:
Hmm... this doesn't seem like the right approach to me.

First, the purpose of the network client configuration is to configure the network client. Currently, we don't need this option; we can use the Infura network name to figure out the appropriate chain ID. Therefore, adding this property expands the responsibility of this type unnecessarily.

Second, this property makes this type ineffective. With this property added, it is now possible to create an InfuraNetworkClientConfiguration that looks like this:

``` typescript
{
  type: NetworkClientType.Infura,
  chainId: "0x1337",
  network: InfuraNetworkType.mainnet,
  infuraProjectId: "infura-project-id"
}
```

That doesn't look right to me. The chain ID associated with an Infura network should never be different than we know it to be and how we construct the network client. If it were, then we'd have a custom network, not an Infura network.

One of the goals of types is to make it impossible for the system to get into an invalid state (even if it is just an accident). This change seems to not do that.

I don't think this property is necessary. I understand that one of our goals is to consolidate the definitions of Infura and custom network clients, but I feel that there is a more cohesive approach we can take.

In this case, we already know what the chain ID for an Infura network is, because we can get it from `BUILT_IN_NETWORKS`. I think a better approach would have been to create a new network client configuration object that combines a custom and an Infura one together so that it now has a common `chainId` and `rpcUrl`. Then converting an Infura network configuration into this common form would look like:

```
{
  chainId: BUILT_IN_NETWORKS.mainnet.chainId,
  rpcUrl: buildInfuraRpcUrl(InfuraNetworkType.mainnet, 'infura-project-id')
}
```

Reply:
An update: @adonesky1, @BelfordZ and @shanejonas and I talked about this. They provided more context behind this change. We need a way to access the chain ID for a network client across many many controllers that we are making multichain capable and this just seemed like the most direct way of solving that constraint without jumping through a bunch of hoops.

Zooming out, this is another case in which the Infura/custom network split is causing difficulties. So of course we know that consolidating the two would solve a lot of problems.

As a more immediate fix, though, we can remove `network` from InfuraNetworkClientConfiguration in favor of `chainId`. In the same way that we can derive chain ID from Infura network name, we can also do the reverse. This should make it possible to access the chain ID of an Infura network client while still preserving the type safety of the two network clients.